cpp/test/cluster/kmeans_balanced.cu

```diff
@@ -0,0 +1,235 @@
+    if constexpr (!std::is_same_v<DataT, MathT>) {
+      raft::linalg::unaryOp(
+        X.data_handle(), blobs.data(), p.n_rows * p.n_cols, op.reverse_op, stream);
+    }
+
+    d_labels.resize(p.n_rows, stream);
+    d_labels_ref.resize(p.n_rows, stream);
+    d_centroids.resize(p.n_clusters * p.n_cols, stream);
+
+    raft::linalg::unaryOp(
+      d_labels_ref.data(), blob_labels.data_handle(), p.n_rows, raft::cast_op<LabelT>(), stream);
+
+    auto X_view =
+      raft::make_device_matrix_view<const DataT, IdxT>(X.data_handle(), X.extent(0), X.extent(1));
+    auto d_centroids_view =
+      raft::make_device_matrix_view<MathT, IdxT>(d_centroids.data(), p.n_clusters, p.n_cols);
+    auto d_labels_view = raft::make_device_vector_view<LabelT, IdxT>(d_labels.data(), p.n_rows);
+
+    raft::cluster::kmeans_balanced::fit_predict(
+      handle, p.kb_params, X_view, d_centroids_view, d_labels_view, op);
+
+    handle.sync_stream(stream);
+
+    score = raft::stats::adjusted_rand_index(
+      d_labels_ref.data(), d_labels.data(), p.n_rows, handle.get_stream());
+
+    if (score < 1.0) {
+      std::stringstream ss;
+      ss << "Expected: " << raft::arr2Str(d_labels_ref.data(), 25, "d_labels_ref", stream);
+      std::cout << (ss.str().c_str()) << '\n';
+      ss.str(std::string());
+      ss << "Actual: " << raft::arr2Str(d_labels.data(), 25, "d_labels", stream);
+      std::cout << (ss.str().c_str()) << '\n';
+      std::cout << "Score = " << score << '\n';
+    }
+  }
+
+  void SetUp() override { basicTest(); }
+
+ protected:
+  raft::handle_t handle;
+  cudaStream_t stream;
+  rmm::device_uvector<LabelT> d_labels;
+  rmm::device_uvector<LabelT> d_labels_ref;
+  rmm::device_uvector<MathT> d_centroids;
+  double score;
+};
+
+template <typename MathT, typename IdxT>
+std::vector<KmeansBalancedInputs<MathT, IdxT>> get_kmeans_balanced_inputs()
+{
+  std::vector<KmeansBalancedInputs<MathT, IdxT>> out;
+  KmeansBalancedInputs<MathT, IdxT> p;
+  p.kb_params.n_iters = 20;
+  p.kb_params.metric  = raft::distance::DistanceType::L2Expanded;
+  p.tol               = MathT{0.0001};
+  std::vector<std::tuple<size_t, size_t, size_t>> row_cols_k = {
+    {1000, 32, 5}, {1000, 100, 20}, {10000, 32, 10}, {10000, 100, 50}, {10000, 500, 100}};
+  for (auto& rck : row_cols_k) {
+    p.n_rows     = static_cast<IdxT>(std::get<0>(rck));
+    p.n_cols     = static_cast<IdxT>(std::get<1>(rck));
+    p.n_clusters = static_cast<IdxT>(std::get<2>(rck));
+    out.push_back(p);
+  }
+  return out;
+}
+
+const auto inputsf_i32 = get_kmeans_balanced_inputs<float, int>();
+const auto inputsd_i32 = get_kmeans_balanced_inputs<double, int>();
+const auto inputsf_i64 = get_kmeans_balanced_inputs<float, int64_t>();
+const auto inputsd_i64 = get_kmeans_balanced_inputs<double, int64_t>();
+
+#define KB_TEST(test_type, test_name, test_inputs)         \
+  typedef RAFT_DEPAREN(test_type) test_name;               \
+  TEST_P(test_name, Result) { ASSERT_TRUE(score == 1.0); } \
```

Comment:
I am not so familiar with the adjusted_rand_score. Is it reasonable to expect score 1.0? If yes, could you add an explicit note about it?

Reply:
That is a discussion we had earlier following failures in the k-means tests. An adjusted rand score of 1.0 means that the clustering must be exactly the same as the reference. It's reasonable to expect this in most cases, but there can be malformed datasets where two blobs are close to each other and therefore some points in the reference might be mislabeled.

If the input generator is deterministic, we don't need to worry about this. But as it is currently not (depends on the GPU), we could potentially run into one of these malformed datasets if we run on a GPU that we haven't run on before. Alternatively, we could fix the blob centers, but it makes it harder to change the test parameters.

Anyway, there is te same problem with the k-means tests, it's not new in this PR, and whatever change we do in one test should be mirrored in the other.